data-prepper-plugins/opensearch/src/main/java/com/amazon/dataprepper/plugins/sink/opensearch/index/IndexManager.java

```diff
@@ -0,0 +1,189 @@
+import org.opensearch.client.indices.IndexTemplatesExistRequest;
+import org.opensearch.client.indices.PutIndexTemplateRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.HttpMethod;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public abstract class IndexManager {
+
+    public static final String INDEX_ALIAS_USED_AS_INDEX_ERROR
+            = "Invalid alias name [%s], an index exists with the same name as the alias";
+    protected RestHighLevelClient restHighLevelClient;
+    protected OpenSearchSinkConfiguration openSearchSinkConfiguration;
+    private static final Logger LOG = LoggerFactory.getLogger(IndexManager.class);
+
+    protected IndexManager(final RestHighLevelClient restHighLevelClient, final OpenSearchSinkConfiguration openSearchSinkConfiguration){
+        checkNotNull(restHighLevelClient);
+        checkNotNull(openSearchSinkConfiguration);
+        this.restHighLevelClient = restHighLevelClient;
+        this.openSearchSinkConfiguration = openSearchSinkConfiguration;
+    }
+
+    public boolean checkISMEnabled() throws IOException {
+        final ClusterGetSettingsRequest request = new ClusterGetSettingsRequest();
+        request.includeDefaults(true);
+        final ClusterGetSettingsResponse response = restHighLevelClient.cluster().getSettings(request, RequestOptions.DEFAULT);
+        final String enabled = response.getSetting(IndexConstants.ISM_ENABLED_SETTING);
+        return enabled != null && enabled.equals("true");
+    }
+
+    public void checkAndCreateIndexTemplate(final boolean isISMEnabled, final String ismPolicyId) throws IOException {
+        final String indexAlias = openSearchSinkConfiguration.getIndexConfiguration().getIndexAlias();
+        final String indexTemplateName = indexAlias + "-index-template";
+
+        // Check existing index template version - only overwrite if version is less than or does not exist
+        if (!shouldCreateIndexTemplate(indexTemplateName)) {
+            return;
+        }
+
+        final PutIndexTemplateRequest putIndexTemplateRequest = new PutIndexTemplateRequest(indexTemplateName);
+
+        putIndexTemplateRequest.patterns(getIndexPatterns(indexAlias));
+
+        if (isISMEnabled) {
+            attachPolicy(openSearchSinkConfiguration.getIndexConfiguration(), ismPolicyId, indexAlias);
+        }
+
+        putIndexTemplateRequest.source(openSearchSinkConfiguration.getIndexConfiguration().getIndexTemplate());
+        restHighLevelClient.indices().putTemplate(putIndexTemplateRequest, RequestOptions.DEFAULT);
+    }
+
+    public abstract Optional<String> checkAndCreatePolicy() throws IOException;
```

Comment:
Thank you for refactoring the Index creation.

I am concerned the proposed architecture does not scale well. The logic for supporting ISM for TraceAnalyticsRawIndex is not unique an generally defines how we should create an ISM index. If we are to add support for ISM as part of log ingestion or custom pipelines we may end up nearly duplicating the code defined in the TraceAnalyticsRawIndexManager with the exceptions of the constants used in that class.

I would like to propose a single index manager that leverages an enum type. The enums can define whether or not they support ISM. 

What do you think?

Reply:
Somewhat along the same lines, I suggested a composition approach.

We could have enums which provide the correct composition as well if we want to continue down the enum path.

Reply:
Good suggestions. Thanks, guys. I have moved ISM related methods to IsmPolicyManagementStrategy: 
```
public interface IsmPolicyManagementStrategy {

    Optional<String> checkAndCreatePolicy() throws IOException;

    List<String> getIndexPatterns(final String indexAlias);

    boolean checkIfIndexExistsOnServer(final String indexAlias) throws IOException;

    CreateIndexRequest getCreateIndexRequest(final String indexAlias);
}


public class IsmPolicyManagement implements IsmPolicyManagementStrategy {.....}

public class NoIsmPolicyManagement implements IsmPolicyManagementStrategy {.......}

```